### DISCUSSION TITLE
Failure:  mvn clean install -e

### DISCUSSION THREAD

**Discussion by profnormano** [wgu-opensource/osmt#108] [Asking for Help]
I saw the same errors when following the Development Stacks at: run ./exportEnv.sh osmt-dev-stack.env  so I followed the alternate approach:

**Alternate approaches to supply env variables.**
you can provide these OAuth2 values as program arguments when starting your Spring Boot app (-Dokta.oauth2.clientId="123456qwerty").
Set environment variables by supplying json
In Windows, replace values and run below command.
set SPRING_APPLICATION_JSON="{\"OAUTH_ISSUER\":\"<aouth_issuer>\",\"OAUTH_CLIENTID\":\"<aouth_client>\", \"OAUTH_CLIENTSECRET\":\"<oauth_clientsecret>\",\"OAUTH_AUDIENCE\":\"<oauth_audience>\"}"
In mac & linux, replace values and run below command.
export SPRING_APPLICATION_JSON="{\"OAUTH_ISSUER\":\"<aouth_issuer>\",\"OAUTH_CLIENTID\":\"<aouth_client>\", \"OAUTH_CLIENTSECRET\":\"<oauth_clientsecret>\",\"OAUTH_AUDIENCE\":\"<oauth_audience>\"}"

[in the above, I correctly inserted the correct values that match my osmt-dev-stack.env file.]

and then I ran "mvn clean install -e"

Errors:

[INFO] Reactor Summary:
[INFO]
[INFO] osmt-parent ........................................ SUCCESS [  0.468 s]
[INFO] osmt-ui ............................................ SUCCESS [01:17 min]
[INFO] osmt-api ........................................... FAILURE [ 21.743 s]
[INFO] ------------------------------------------------------------------------

...

[ERROR] Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.3.72:compile (compile) on project osmt-api: Compilation failure: Compilation failure:
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt:[83,36] Unresolved reference: strip
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt:[84,32] Unresolved reference: strip
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt:[85,44] Unresolved reference: strip
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.3.72:compile (compile) on project osmt-api: Compilation failure

**Comment by JohnKallies** [comment]
I've never used that approach, a colleague added it to the docs. JSON requires double quotes. If you echo $SPRING_APPLICATION_JSON, did everything escape OK? If you drop that value in a pretty printer/validator, the JSON is good?

**Comment by profnormano** [comment]
The alternate approach worked just fine.   I tried it after I received the error on "mvn clean install -e" in case there was a possibility of the error being related to environments.

Any thoughts on the compilation error?

**Comment by JohnKallies** [comment]
To confirm, you are building on an EC2, using Maven, and not in the context of a Docker build for a Spring app image. Do I have this correct?

Assuming yes, the Kotlin version is managed by the Maven build. What do you get for `mvn --version`?

**Comment by profnormano** [comment]
To the best of my ability I am following the README which includes the Docker build.   So in "Development Configuration" I follow Step 1 for Docker and then move to Step 2.

mvn --version
Apache Maven 3.5.2 (138edd61fd100ec658bfa2d307c43b76940a5d7d; 2017-10-18T07:58:13Z)
Maven home: /usr/share/apache-maven
Java version: 1.8.0_312, vendor: Red Hat, Inc.
Java home: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.312.b07-1.amzn2.0.2.x86_64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.10.82-83.359.amzn2.x86_64", arch: "amd64", family: "unix"

**Reply by JohnKallies** [reply]
Ok. Are you able to use Java 11? 

The project has 3 different Dockerfiles

```
> find . -name "*Dockerfile*"
./docker/base-images/Dockerfile.osmt-build
./docker/base-images/Dockerfile.osmt-base
./Dockerfile
```

- The "base" is a common CentOS 8, stripped of old Javas, and OpenJDK 11 installed
- The "build" image extends the "base" image, and has Maven 3.8.4 installed
- The one in the project root is for the OSMT app. It extends the "base" image, and has the app jar and docker-entrypoint.sh added.

I don't know of anyone running OSMT with Java 8. The docs call out Java 11 as a baseline. Can you use the Dockerfiles to adjust your RedHat image, or select a RedHat with Java 11 already installed?

**Reply by profnormano** [reply]
Doh!  I'll do that install and report back.  Thank-you!

**Comment by profnormano** [comment]
Same error after this command:  ./exportEnv.sh osmt-dev-stack.env
versions called out at the bottom.

[INFO] --- kotlin-maven-plugin:1.3.72:compile (compile) @ osmt-api ---
[INFO] Applied plugin: 'spring'
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt: (83, 36) Unresolved reference: strip
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt: (84, 32) Unresolved reference: strip
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt: (85, 44) Unresolved reference: strip
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 27.033 s
[INFO] Finished at: 2022-01-04T21:21:53Z
[INFO] Final Memory: 129M/671M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.3.72:compile (compile) on project osmt-api: Compilation failure: Compilation failure:
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt:[83,36] Unresolved reference: strip
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt:[84,32] Unresolved reference: strip
[ERROR] /home/ec2-user/osmt/api/src/main/kotlin/edu/wgu/osmt/keyword/KeywordRepository.kt:[85,44] Unresolved reference: strip
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ec2-user@ip-10-0-0-167 api]$ mvn --version
Apache Maven 3.5.2 (138edd61fd100ec658bfa2d307c43b76940a5d7d; 2017-10-18T07:58:13Z)
Maven home: /usr/share/apache-maven
Java version: 1.8.0_312, vendor: Red Hat, Inc.
Java home: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.312.b07-1.amzn2.0.2.x86_64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.10.82-83.359.amzn2.x86_64", arch: "amd64", family: "unix"
[ec2-user@ip-10-0-0-167 api]$ java -version
openjdk version "11.0.13" 2021-10-19 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.13+8-LTS, mixed mode, sharing)
[ec2-user@ip-10-0-0-167 api]$ node --version
v17.3.0
[ec2-user@ip-10-0-0-167 api]$ npm -version
8.3.0

**Comment by profnormano** [comment]
I'll try a rebuild from a new EC2 today.  It will also help me refine my documentation for an EC2 install.

**Comment by JohnKallies** [comment]
If this is the output from `mvn --version`, then this instance is at least using a Java 8.

> Java version: 1.8.0_312, vendor: Red Hat, Inc.
> Java home: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.312.b07-1.amzn2.0.2.x86_64/jre

There may be a Java 11 available, but in this context, it's not using it.

Is there an AMI with your favored OS and Java 11 ready to go?

Also... it seems to me you are working to run the Spring app as a jar on an EC2, and you are choosing to **not** run it as a Docker container. Can you please confirm that this is your intention?
